marshal.go

```diff
@@ -0,0 +1,272 @@
+	}
+	p := &Params{req, httprouter.Params{}}
+	if err := marshal(p, xv, pt); err != nil {
+		return nil, errgo.Mask(err, errgo.Is(ErrUnmarshal))
+	}
+	return p.Request, nil
+}
+
+// marshal is the internal version of Marshal.
+func marshal(p *Params, xv reflect.Value, pt *requestType) error {
+	if xv.Kind() == reflect.Ptr {
+		xv = xv.Elem()
+	}
+	for _, f := range pt.fields {
+		fv := xv.FieldByIndex(f.index)
+
+		// TODO store the field name in the field so
+		// that we can produce a nice error message.
+		if err := f.marshal(fv, p); err != nil {
+			return errgo.WithCausef(err, ErrUnmarshal, "cannot marshal field")
+		}
+	}
+
+	urlString := p.URL.Path
+	var pathBuffer bytes.Buffer
+	paramsByName := make(map[string]string)
+	for _, param := range p.PathVar {
+		paramsByName[param.Key] = param.Value
+	}
+
+	offset := 0
+	hasParams := false
+	for i := 0; i < len(urlString); i++ {
+		c := urlString[i]
+		if c != ':' {
+			continue
+		}
+		hasParams = true
+
+		end := i + 1
+		for end < len(urlString) && urlString[end] != ':' && urlString[end] != '/' {
+			end++
+		}
+
+		if end-i < 2 {
+			return errgo.New("request wildcards must be named with a non-empty name")
+		}
+		if i > 0 {
+			pathBuffer.WriteString(urlString[offset:i])
+		}
+
+		wildcard := urlString[i+1 : end]
+		paramValue, ok := paramsByName[wildcard]
+		if !ok {
+			return errgo.Newf("missing value for path parameter %q", wildcard)
+		}
+		pathBuffer.WriteString(paramValue)
+		offset = end
+	}
+	if !hasParams {
+		pathBuffer.WriteString(urlString)
+	}
+
+	p.URL.Path = pathBuffer.String()
+
+	p.URL.RawQuery = p.Form.Encode()
+
+	return nil
+}
+
+// getMarshaler returns a marshaler function suitable for marshaling
+// a field with the given tag into and http request.
+func getMarshaler(tag tag, t reflect.Type) (marshaler, error) {
+	switch {
+	case tag.source == sourceNone:
+		return marshalNop, nil
+	case tag.source == sourceBody:
+		return marshalBody, nil
+	case t == reflect.TypeOf([]string(nil)):
+		if tag.source != sourceForm {
+			return nil, errgo.New("invalid target type []string for path parameter")
+		}
+		return marshalAllField(tag.name), nil
+	case t == reflect.TypeOf(""):
+		return marshalString(tag), nil
+	case implementsTextMarshaler(t):
+		return marshalWithMarshalText(t, tag), nil
+	default:
+		return marshalWithSprint(tag), nil
+	}
+}
+
+func dereferenceIfPointer(v reflect.Value) reflect.Value {
+	if v.Kind() == reflect.Ptr {
+		if v.IsNil() {
+			return reflect.Value{}
+		}
+		return v.Elem()
+	}
+	return v
+}
+
+// marshalNop does nothing with the value.
+func marshalNop(v reflect.Value, p *Params) error {
+	return nil
+}
+
+// mashalBody marshals the specified value into the body of the http request.
+func marshalBody(v reflect.Value, p *Params) error {
+	// TODO allow body types that aren't necessarily JSON.
+	bodyValue := dereferenceIfPointer(v)
+	if bodyValue == emptyValue {
+		return nil
+	}
+
+	if p.Method != "POST" && p.Method != "PUT" {
```

Comment:
I think it would probably be better to exclude known bad methods than
only allow good ones here, because it's allowed to define new
methods. For example, PATCH would be just fine here.

Perhaps just exclude GET and HEAD as being the most
likely common culprits?